cpp/include/cudf/ast/detail/transform.cuh

```diff
@@ -40,132 +42,375 @@ namespace ast {
   {
-    cudf_assert(false && "Invalid type in resolve_output.");
+    subclass()->set_value();
   }
 
- private:
-  row_evaluator const& evaluator;
+  __device__ bool is_valid() const { subclass()->is_valid(); }
+
+  __device__ T value() const { subclass()->value(); }
 };
 
-template <typename Input>
-struct unary_row_output : public row_output {
-  __device__ unary_row_output(row_evaluator const& evaluator) : row_output(evaluator) {}
+/**
+ * @brief A container for capturing the output of an evaluated expression in a scalar.
+ *
+ * This subclass of `expression_result` functions as an owning container of a
+ * (possibly nullable) scalar type that can be written to by the
+ * expression_evaluator. The data (and its validity) can then be accessed.
+ *
+ * @tparam T The underlying data type.
+ * @tparam has_nulls Whether or not the result data is nullable.
+ */
+template <typename T, bool has_nulls>
+struct value_expression_result
+  : public expression_result<value_expression_result<T, has_nulls>, T, has_nulls> {
+  __device__ value_expression_result() {}
 
-  template <
-    ast_operator op,
-    std::enable_if_t<detail::is_valid_unary_op<detail::operator_functor<op>, Input>>* = nullptr>
-  __device__ void operator()(cudf::size_type row_index,
-                             Input input,
-                             detail::device_data_reference output) const
+  template <typename Element>
+  __device__ void set_value(cudf::size_type index, possibly_null_value_t<Element, has_nulls> result)
   {
-    using OperatorFunctor = detail::operator_functor<op>;
-    using Out             = cuda::std::invoke_result_t<OperatorFunctor, Input>;
-    resolve_output<Out>(output, row_index, OperatorFunctor{}(input));
+    if constexpr (std::is_same_v<Element, T>) {
+      _obj = result;
+    } else {
+      cudf_assert(false && "Output type does not match container type.");
+    }
   }
 
-  template <
-    ast_operator op,
-    std::enable_if_t<!detail::is_valid_unary_op<detail::operator_functor<op>, Input>>* = nullptr>
-  __device__ void operator()(cudf::size_type row_index,
-                             Input input,
-                             detail::device_data_reference output) const
+  /**
+   * @brief Returns true if the underlying data is valid and false otherwise.
+   */
+  __device__ bool is_valid() const
   {
-    cudf_assert(false && "Invalid unary dispatch operator for the provided input.");
+    if constexpr (has_nulls) { return _obj.has_value(); }
+    return true;
   }
+
+  /**
+   * @brief Returns the underlying data.
+   *
+   * @throws thrust::bad_optional_access if the underlying data is not valid.
+   */
+  __device__ T value() const
+  {
+    // Using two separate constexprs silences compiler warnings, whereas an
+    // if/else does not. An unconditional return is not ignored by the compiler
+    // when has_nulls is true and therefore raises a compiler error.
+    if constexpr (has_nulls) { return _obj.value(); }
```

Comment:
`optional::value` performs error checking and throws if `has_value` is false. This is obviously problematic in device code and causes UB (nvcc ignores the `throw` because of `--expt-relaxed-constexpr` see https://github.com/rapidsai/cudf/issues/7795). Use `operator*` instead to not perform an error checking. 
```suggestion
    if constexpr (has_nulls) { return *_obj; }
```

Reply:
Ah of course good catch.